### PR TITLE
Restructured mul_dense_dense() function

### PR DESCRIPTION
**Pull request by ShikharJ**
@certik @isuruf I would be requiring some clarification for the second part of the issue. Please check my comment: https://github.com/symengine/symengine/issues/1025#issuecomment-265145254

### CODE REVIEW

**Comment by ShikharJ** [comment]
ping @isuruf @certik

**Comment by certik** [comment]
I am +0, I don't know if this is needed, I would rather do these checks at Debug time only. I'll leave this to @isuruf to decide.

**Comment by ShikharJ** [comment]
@isuruf I'm frankly out of ideas regarding the second part of the issue. The alternative that appears to me is using a temporary like this:
```
void mul_dense_dense(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
{
    SYMENGINE_ASSERT(A.col_ == B.row_);

    DenseMatrix tmp = DenseMatrix(A.row_, B.col_);

    unsigned row = A.row_, col = B.col_;

    for (unsigned r = 0; r < row; r++) {
        for (unsigned c = 0; c < col; c++) {
            tmp.m_[r * col + c] = zero; // Integer Zero
            for (unsigned k = 0; k < A.col_; k++)
                tmp.m_[r * col + c]
                        = add(tmp.m_[r * col + c],
                              mul(A.m_[r * A.col_ + k], B.m_[k * col + c]));
        }
    }

    C = tmp;
}
```

**Comment by ShikharJ** [comment]
Ping @isuruf.

**Comment by isuruf** [comment]
@ShikharJ, using a temporary is fine, but only if necessary. If C is neither A nor B, it's better to just use C

**Comment by ShikharJ** [comment]
@isuruf Is there any way if they are the same object? Using `A != C` or `B != C` doesn't seem to work.

**Comment by isuruf** [comment]
`&A != &C` ?

**Comment by ShikharJ** [comment]
Updated @isuruf.

File: symengine/dense_matrix.cpp [hunk 1]
```diff
@@ -327,19 +327,27 @@ void add_dense_scalar(const DenseMatrix &A, const RCP<const Basic> &k,
 // ------------------------------- Matrix Multiplication ---------------------//
 void mul_dense_dense(const DenseMatrix &A, const DenseMatrix &B, DenseMatrix &C)
 {
-    SYMENGINE_ASSERT(A.col_ == B.row_ and C.row_ == A.row_
-                     and C.col_ == B.col_);
+    SYMENGINE_ASSERT(A.col_ == B.row_);
+
+    if (C.row_ != A.row_ or C.col_ != B.col_)
+        C.resize(A.row_, B.col_);
```
Comment (isuruf) [inline symengine/dense_matrix.cpp]: If you do this and `C` and `A` are the same matrix, changing `C` might affect `A`

Comment (isuruf) [inline symengine/dense_matrix.cpp]: Since we don't have a clear consensus, let's remove the changes above and keep the changes below

**Comment by ShikharJ** [comment]
Updated @isuruf.

**Pull request merged**